app/Http/Controllers/Auth/ServerController.php

```diff
@@ -16,19 +16,53 @@ class ServerController extends Controller
 {
     public function beginLogin(Request $request)
     {
-        $client_token = $request->input('token', '');
-        // Check if out token is MD5 hash, what it should be
-        if(!preg_match('/^[a-f0-9]{32}$/', $client_token)) {
+        $client_token = 'Inv@lid';
+        if($request->has('token'))
+        {
+            $client_token = $request->input('token');
+            // Check if out token is MD5 hash, what it should be
+            if(!preg_match('/^[a-f0-9]{32}$/', $client_token)) {
+                abort(400, 'Invalid request.');
+            }
+            // Let's store our token for later use
+            $request->session()->put('server_client_token', $client_token);
+        } elseif ($request->session()->has('server_client_token')) {
+            // If we already have token, aka, we are being redirected, then we just use that token.
+            $client_token = $request->session()->pull('server_client_token');
+        } else {
             abort(400, 'Invalid request.');
         }
-        // Let's store our token for later use
-        $request->session()->put('server_client_token', $client_token);
+
         if (!Auth::check()) {
             return redirect('login');
         }
+
+        $query = New ServerQuery;
+        try {
+            $query->setUp(config('aurora.gameserver_address'),config('aurora.gameserver_port'),config('aurora.gameserver_auth'));
+            $query->runQuery([
+                'query' => 'get_auth_client_ip',
+                'clienttoken' => $client_token,
+            ]);
+        } catch (\Exception $e) {
+            abort(500, $e->getMessage());
+        }
+        if ($query->response->statuscode != '200') {
+            abort($query->response->statuscode);
+        }
+
+        if($request->getClientIp() !== $query->response->data) {
+            redirect()->route('server.login.warn');
```

Comment:
While I am unsure of the structure here. Shouldn't this `return redirect()->route('server.login.warn');` ?